src/session/session.c

```diff
@@ -400,15 +399,15 @@ exit_init_problem (int result_code)
   else
     message = pam_strerror (NULL, result_code);
 
-  if (asprintf (&payload, "\n{\"command\":\"init\",\"version\":1,\"problem\":\"%s\",\"message\":\"%s\"}",
-                problem, message) < 0)
-    errx (EX, "couldn't allocate memory for message");
+  /* we send the message through the Cockpit protocol, make sure we don't break JSON encoding */
+  char *message_sanitized = strdupx (message);
+  char *c;
+  while ((c = strchr (message_sanitized, '"')) != NULL)
```

Comment:
Didn't look close enough the first time around (because of the more obvious bug) but my understanding of this code wasn't correct: you're rescanning the *entire* string on every loop iteration, not just starting again from the point of `c`.

This is valid though, because the previous "match" will no longer exist, since you've replaced it.